doc/designs/evp-cipher-pipeline.md

```diff
@@ -0,0 +1,175 @@
+**EVP Interface:**
+API to check for pipeline support in provided cipher.
+```c
+/**
+ * @brief checks if the provider has exported required pipeline functions
+ * @return 0 (pipeline not supported) or 1 (pipeline supported)
+ */
+int EVP_CIPHER_can_pipeline(const EVP_CIPHER *cipher)
+```
+
+Multi-call APIs for init, update and final. Associated data for AEAD ciphers are set in `EVP_CipherPipelineUpdate`.
+```c
+/**
+ * @param iv    array of pointers (array length must be numpipes)
+ */
+EVP_CipherPipelineInit(EVP_CIPHER_CTX *ctx, const EVP_CIPHER *cipher, const unsigned char *key, size_t numpipes, const unsigned char **iv, int enc);
+
+/**
+ * @param out   array of pointers to output buffers (array length must be numpipes)
+ *              when NULL, input buffers are treated as AAD data
+ * @param outl  array of size_t (array length must be numpipes)
+ * @param in    array of pointers to input buffers (array length must be numpipes)
+ * @param inl   array of size_t (array length must be numpipes)
+ */
+EVP_CipherPipelineUpdate(EVP_CIPHER_CTX *ctx, unsigned char **out, size_t *outl, const unsigned char **in, size_t *inl);
+
+/**
+ * @param outm  array of pointers to output buffers (array length must be numpipes)
+ * @param outl  array of size_t (array length must be numpipes)
+ */
+EVP_CipherPipelineFinal(EVP_CIPHER_CTX *ctx, unsigned char **outm, size_t *outl);
+```
+
+API to get/set AEAD auth tag.
+```c
+/**
+ * @param buf   array of pointers to in/out buffers (array length must be numpipes)
+ * @param bsize aead tag length
+ */
+OSSL_CIPHER_PARAM_PIPELINE_AEAD_TAG (type OSSL_PARAM_OCTET_PTR)
+```
+
+**Design Decisions:**
+1. Denoting pipeline support
+    - [ ] a. A cipher flag `EVP_CIPH_FLAG_PROVIDED_PIPELINE` (this has to be different than EVP_CIPH_FLAG_PIPELINE, so that it doesn't break legacy applications).
+    - [x] b. A function `EVP_CIPHER_can_pipeline()` that checks if the provider exports pipeline functions.
+    > **Justification:** flags variable is deprecated in EVP_CIPHER struct. Moreover, EVP can check for presence of pipeline functions, rather than requiring providers to set a flag.
+
+2. `numpipes` argument
+    - [x] a. `numpipes` received only in `EVP_CipherPipelineInit()` and saved in EVP_CIPHER_CTX for further use.
+    - [ ] b. `numpipes` value is repeatedly received in each `EVP_CipherPipelineInit()`, `EVP_CipherPipelineUpdate()` and `EVP_CipherPipelineFinal()` call.
+    > **Justification:** It is expected for numpipes to be same across init, update and final operation.
+
+3. Input/Output buffers
+    - [ ] a. A set of buffers is represented by an array of buffer pointers and an array of lengths. Example: `unsigned char **out, size_t *outl`.
+    - [ ] b. iovec style: A new type that holds one buffer pointer along with its size.
+    ```c
+    struct {
+        unsigned char *buf;
+        size_t buf_len;
+    } iovec_buf;
+    EVP_CipherPipelineUpdate(EVP_CIPHER_CTX *ctx, struct iovec_buf in, struct iovec_buf *out);
```

Comment:
I have included the stride argument in design. However, this may require reconsideration. The stride argument is only essential if there's a possibility of the struct changing. In this instance, it's a struct designed to represent a C buffer. Beyond the pointer and size, I cannot foresee any additional meaningful fields.